### ISSUE TITLE
Grove tests break on Python 3.6

### ISSUE THREAD

**Issue by jotterbach** [rigetticomputing/grove#106]
running tox on python 3.6 leads to errors due to missing hasability of pyQuil programs in the latest version of pyQuil:

```____________________________________________________________________________________________________ test_x_oracle_one_grover _____________________________________________________________________________________________________

x_oracle = (<pyquil.quil.Program object at 0x10a839a90>, <Qubit 10A839BE0>)

    def test_x_oracle_one_grover(x_oracle):
        """Testing that Grover's algorithm with an oracle that applies an X gate to the query bit works,
         with one iteration."""
        x_oracle_grover = Program()
        qubit0 = x_oracle_grover.alloc()
        qubits = [qubit0]
        oracle, query_qubit = x_oracle
        with patch("pyquil.quilbase.InstructionGroup.alloc") as mock_alloc:
            mock_alloc.return_value = qubit0
        generated_x_oracle_grover = Grover().oracle_grover(oracle, qubits, 1)
        # First we put the input into uniform superposition.
        x_oracle_grover.inst(H(qubit0))
        # Now an oracle is applied.
        x_oracle_grover.inst(X(query_qubit))
        # We now apply the diffusion operator.
        x_oracle_grover.inst(H(qubit0))
        x_oracle_grover.inst(Z(qubit0))
        x_oracle_grover.inst(H(qubit0))
>       synthesize_programs(generated_x_oracle_grover, x_oracle_grover)

grove/tests/amplification/test_grover.py:60: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

progs = (<pyquil.quil.Program object at 0x10a85ebe0>, <pyquil.quil.Program object at 0x10a8306a0>), get_allocated_qubits = <function synthesize_programs.<locals>.get_allocated_qubits at 0x10a83a378>
qubits_to_progs = {<Qubit 10A839BE0>: [<pyquil.quil.Program object at 0x10a85ebe0>, <pyquil.quil.Program object at 0x10a85ebe0>, <pyquil...object at 0x10a85ebe0>, <pyquil.quil.Program object at 0x10a85ebe0>, <pyquil.quil.Program object at 0x10a8306a0>, ...]}
prog = <pyquil.quil.Program object at 0x10a8306a0>

    def synthesize_programs(*progs):
        """Synthesizes programs together, paying attention to shared qubits. This breaks if synthesize
         is called after, very fragile.
    
        :param progs: List of Programs.
        :return: None
        :rtype: NoneType"""
        def get_allocated_qubits(*progs):
            qubits_to_progs = {}
            for prog in progs:
                for inst in prog:
                    if inst[0] == ACTION_INSTANTIATE_QUBIT:
                        qubits_to_progs[inst[1]] = [prog]
            return qubits_to_progs
        qubits_to_progs = get_allocated_qubits(*progs)
        for prog in progs:
            for inst in prog:
                if isinstance(inst[1], Gate):
                    for arg in inst[1].arguments:
                        if isinstance(arg, Qubit):
                            qubits_to_progs[arg].append(prog)
    
        equiv_classes = list(qubits_to_progs.values())
        repeat = True
        while repeat:
            repeat = False
            for i, equiv_class in enumerate(equiv_classes):
                for j, other_equiv_class in enumerate(equiv_classes):
                    if j <= i:
                        continue
>                   if set(equiv_class).intersection(set(other_equiv_class)):
E                   TypeError: unhashable type: 'Program'

grove/tests/utils/utils_for_testing.py:78: TypeError
____________________________________________________________________________________________________ test_x_oracle_two_grover _____________________________________________________________________________________________________

x_oracle = (<pyquil.quil.Program object at 0x108637d30>, <Qubit 108637BE0>)

    def test_x_oracle_two_grover(x_oracle):
        """Testing that Grover's algorithm with an oracle that applies an X gate to the query bit works,
         with two iterations."""
        x_oracle_grover = Program()
        qubit0 = x_oracle_grover.alloc()
        qubits = [qubit0]
        oracle, query_qubit = x_oracle
        with patch("pyquil.quilbase.InstructionGroup.alloc") as mock_alloc:
            mock_alloc.return_value = qubit0
        generated_x_oracle_grover = Grover().oracle_grover(oracle, qubits, 2)
        # First we put the input into uniform superposition.
        x_oracle_grover.inst(H(qubit0))
        # Two iterations.
        for _ in range(2):
            # Now an oracle is applied.
            x_oracle_grover.inst(X(query_qubit))
            # We apply the diffusion operator.
            x_oracle_grover.inst(H(qubit0))
            x_oracle_grover.inst(Z(qubit0))
            x_oracle_grover.inst(H(qubit0))
>       synthesize_programs(generated_x_oracle_grover, x_oracle_grover)

grove/tests/amplification/test_grover.py:84: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

progs = (<pyquil.quil.Program object at 0x10a89ff60>, <pyquil.quil.Program object at 0x10a818b70>), get_allocated_qubits = <function synthesize_programs.<locals>.get_allocated_qubits at 0x10a87cb70>
qubits_to_progs = {<Qubit 108637BE0>: [<pyquil.quil.Program object at 0x10a89ff60>, <pyquil.quil.Program object at 0x10a89ff60>, <pyquil...object at 0x10a89ff60>, <pyquil.quil.Program object at 0x10a89ff60>, <pyquil.quil.Program object at 0x10a89ff60>, ...]}
prog = <pyquil.quil.Program object at 0x10a818b70>

    def synthesize_programs(*progs):
        """Synthesizes programs together, paying attention to shared qubits. This breaks if synthesize
         is called after, very fragile.
    
        :param progs: List of Programs.
        :return: None
        :rtype: NoneType"""
        def get_allocated_qubits(*progs):
            qubits_to_progs = {}
            for prog in progs:
                for inst in prog:
                    if inst[0] == ACTION_INSTANTIATE_QUBIT:
                        qubits_to_progs[inst[1]] = [prog]
            return qubits_to_progs
        qubits_to_progs = get_allocated_qubits(*progs)
        for prog in progs:
            for inst in prog:
                if isinstance(inst[1], Gate):
                    for arg in inst[1].arguments:
                        if isinstance(arg, Qubit):
                            qubits_to_progs[arg].append(prog)
    
        equiv_classes = list(qubits_to_progs.values())
        repeat = True
        while repeat:
            repeat = False
            for i, equiv_class in enumerate(equiv_classes):
                for j, other_equiv_class in enumerate(equiv_classes):
                    if j <= i:
                        continue
>                   if set(equiv_class).intersection(set(other_equiv_class)):
E                   TypeError: unhashable type: 'Program'

grove/tests/utils/utils_for_testing.py:78: TypeError
_______________________________________________________________________________________________________ test_optimal_grover _______________________________________________________________________________________________________

x_oracle = (<pyquil.quil.Program object at 0x10a8b1208>, <Qubit 10A8B1278>)

    def test_optimal_grover(x_oracle):
        """Testing that Grover's algorithm with an oracle that applies an X gate to the query bit works,
         and defaults to the optimal number of iterations."""
        grover_precircuit = Program()
        qubit0 = grover_precircuit.alloc()
        qubits = [qubit0]
        oracle, query_qubit = x_oracle
        with patch("pyquil.quilbase.InstructionGroup.alloc") as mock_alloc:
            mock_alloc.return_value = qubit0
            generated_one_iter_grover = Grover().oracle_grover(oracle, qubits)
        # First we put the input into uniform superposition.
        grover_precircuit.inst(H(qubit0))
        # We only do one iteration, which is the result of rounding pi * sqrt(N)/4
        iter = Program()
    
        # An oracle is applied.
        iter.inst(X(query_qubit))
        # We now apply the diffusion operator.
        iter.inst(H(qubit0))
        iter.inst(Z(qubit0))
        iter.inst(H(qubit0))
        one_iter_grover = grover_precircuit + iter
>       synthesize_programs(generated_one_iter_grover, one_iter_grover)

grove/tests/amplification/test_grover.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

progs = (<pyquil.quil.Program object at 0x10a89a780>, <pyquil.quil.Program object at 0x10a89a208>), get_allocated_qubits = <function synthesize_programs.<locals>.get_allocated_qubits at 0x10a87ca60>
qubits_to_progs = {<Qubit 10A8B1278>: [<pyquil.quil.Program object at 0x10a89a780>, <pyquil.quil.Program object at 0x10a89a780>, <pyquil...object at 0x10a89a780>, <pyquil.quil.Program object at 0x10a89a780>, <pyquil.quil.Program object at 0x10a89a208>, ...]}
prog = <pyquil.quil.Program object at 0x10a89a208>

    def synthesize_programs(*progs):
        """Synthesizes programs together, paying attention to shared qubits. This breaks if synthesize
         is called after, very fragile.
    
        :param progs: List of Programs.
        :return: None
        :rtype: NoneType"""
        def get_allocated_qubits(*progs):
            qubits_to_progs = {}
            for prog in progs:
                for inst in prog:
                    if inst[0] == ACTION_INSTANTIATE_QUBIT:
                        qubits_to_progs[inst[1]] = [prog]
            return qubits_to_progs
        qubits_to_progs = get_allocated_qubits(*progs)
        for prog in progs:
            for inst in prog:
                if isinstance(inst[1], Gate):
                    for arg in inst[1].arguments:
                        if isinstance(arg, Qubit):
                            qubits_to_progs[arg].append(prog)
    
        equiv_classes = list(qubits_to_progs.values())
        repeat = True
        while repeat:
            repeat = False
            for i, equiv_class in enumerate(equiv_classes):
                for j, other_equiv_class in enumerate(equiv_classes):
                    if j <= i:
                        continue
>                   if set(equiv_class).intersection(set(other_equiv_class)):
E                   TypeError: unhashable type: 'Program'

grove/tests/utils/utils_for_testing.py:78: TypeError
========================================================================================= 3 failed, 82 passed, 10 skipped in 4.29 seconds =========================================================================================
ERROR: InvocationError: '/Users/johannes/code/grove/.tox/py36/bin/py.test --cov=grove'
_____________________________________________________________________________________________________________ summary _____________________________________________________________________________________________________________
ERROR:   py36: commands failed
```

**Comment by jotterbach** [comment]
This will be fixed in the upcoming release of pyQuil, see https://github.com/rigetticomputing/pyquil/pull/180
cc @stevenheidel

**Comment by jotterbach** [comment]
We fixed this with the latest release of v1.4.0
